storage/raftstorage.go

```diff
@@ -0,0 +1,376 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Ben Darnell
+
+package storage
+
+import (
+	"sync/atomic"
+
+	"github.com/cockroachdb/cockroach/multiraft"
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/storage/engine"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/encoding"
+	"github.com/coreos/etcd/raft"
+	"github.com/coreos/etcd/raft/raftpb"
+	gogoproto "github.com/gogo/protobuf/proto"
+)
+
+var _ multiraft.WriteableGroupStorage = &Range{}
+
+// InitialState implements the raft.Storage interface.
+func (r *Range) InitialState() (raftpb.HardState, raftpb.ConfState, error) {
+	var hs raftpb.HardState
+	found, err := engine.MVCCGetProto(r.rm.Engine(), engine.RaftHardStateKey(r.Desc().RaftID),
+		proto.ZeroTimestamp, true, nil, &hs)
+	if err != nil {
+		return raftpb.HardState{}, raftpb.ConfState{}, err
+	}
+	if !found {
+		// We don't have a saved HardState, so set up the defaults.
+		if r.isInitialized() {
+			// Set the initial log term.
+			hs.Term = raftInitialLogTerm
+			hs.Commit = raftInitialLogIndex
+
+			atomic.StoreUint64(&r.lastIndex, raftInitialLogIndex)
+		} else {
+			// This is a new range we are receiving from another node. Start
+			// from zero so we will receive a snapshot.
+			atomic.StoreUint64(&r.lastIndex, 0)
+		}
+	}
+
+	var cs raftpb.ConfState
+	// For uninitalized ranges, membership is unknown at this point.
+	if found || r.isInitialized() {
+		for _, rep := range r.Desc().Replicas {
+			cs.Nodes = append(cs.Nodes, uint64(MakeRaftNodeID(rep.NodeID, rep.StoreID)))
+		}
+	}
+
+	return hs, cs, nil
+}
+
+// loadLastIndex looks in the engine to find the last log index.
+func (r *Range) loadLastIndex() error {
+	lastIndex := uint64(0)
+	raftID := r.Desc().RaftID
+	eng := r.rm.Engine()
+	v, err := engine.MVCCGet(eng, engine.RaftLastIndexKey(raftID),
+		proto.ZeroTimestamp, true, nil)
+	if err != nil {
+		return err
+	}
+	if v != nil {
+		_, lastIndex = encoding.DecodeUint64(v.Bytes)
+	} else {
```

Comment:
I think this part needs an update. Probably Truncate needs to make sure to kill the kv pair that now persists the last index. Entries will have to do the same.